README.md
Astarte Stream Qt5 Test
=======================

Astarte Qt5 SDK based data stream test.

Requirements
============

* Astarte Qt5 SDK and Qt5 devel tools

Getting started
===============

## Choosing a Device ID

A Base64 url-encoded uuid should be used, you can use [astartectl](https://github.com/astarte-platform/astartectl#installation) to generate one:

```
astartectl utils device-id generate-random
```

## Configuring the application

Create a configuration directory using the given template ([astarte-device-DEVICE_ID_HERE-conf](https://github.com/astarte-platform/stream-qt5-test/tree/release-1.0/astarte-device-DEVICE_ID_HERE-conf)), substituting the previously generated device ID to `DEVICE_ID_HERE`.
Make sure to have your realm's private key at hand.
Then customize the following fields:

- `pairingJwt`: If you don't already have an Astarte pairing JWT, you can generate one with:
`
astartectl utils gen-jwt pairing -k <realm_private_key>
`
- `pairingUrl`: the URL of the Astarte Pairing endpoint. It should be something like `https://<api url>/pairing`.
- `realm`: the name of the Astarte realm.
- `persistencyDir`: When using Docker, allows to save the container's persistency through a Docker Volume.


## Build and run

Build the application using following commands:
```
$ qmake .
$ make
```
Then run the application with:
```
./stream-qt5-test --device <your device ID>
```
Several options are available:

- `--device` allows to set the device ID;
- `--function` allows to choose the data generation function (one between `sin`, `noisesin`, `randomspikessin`, `saw`, `rect`, `sinc`, `random`, `x` and a default one);
- `--interval` allows to set the sending interval;
- `--scale` allows to scale the generated result;
- `--interface` allows to specify the Astarte interface installed on the device;
- `--path` allows to specify the mapping endpoint for the interface.

## Configuring the streaming simulator via Astarte Interface

Configuring the streaming options [`function`, `interval`, `scale`] can be accomplished by sending some data from Astarte to the device via `org.astarte-platform.config.Telemetry` interface. This can be done using the Astarte AppEngine API or with astartectl command.
### `Sin` example:

- Enable the `sin` data generation function, with default period 
``` shell
astartectl appengine --appengine-url https://<api url>/appengine    \
 --token $APPENGINE_TOKEN --realm-name $REALM_NAME                  \ 
 devices send-data $DEVICE_ID                                       \
 org.astarte-platform.config.Telemetry                           \
 --interface-type properties --payload-type boolean                 \
 /request/sin/enable true  
```

- Change the `sin` sending interval to 1000ms
``` shell
astartectl appengine --appengine-url https://<api url>/appengine    \
 --token $APPENGINE_TOKEN --realm-name $REALM_NAME                  \ 
 devices send-data $DEVICE_ID                                       \
 org.astarte-platform.config.Telemetry                           \
 --interface-type properties --payload-type integer                 \
 /request/sin/periodMS 1000  
```

- Change the `sin` scale to 0.0001
``` shell
astartectl appengine --appengine-url https://<api url>/appengine    \
 --token $APPENGINE_TOKEN --realm-name $REALM_NAME                  \ 
 devices send-data $DEVICE_ID                                       \
 org.astarte-platform.config.Telemetry                           \
 --interface-type properties --payload-type double                 \
 /request/sin/scale 0.0001
```

Using the Docker image
====

Make sure to have [astartectl](https://github.com/astarte-platform/astartectl#installation) installed and your realm's private key at hand.

Pull the image with:
```
docker pull astarte/astarte-stream-qt5-test:1.0.4
```

If you are running the container on a different machine than your Astarte instance, you can invoke the application with:
```
docker run -e "DEVICE_ID=$(astartectl utils device-id generate-random)" -e "PAIRING_URL=<your pairing url>" -e "REALM=<realm name>" -e "PAIRING_JWT=$(astartectl utils gen-jwt pairing -k <realm private key>)" -e astarte/astarte-stream-qt5-test:1.0.4
```

Otherwise, to make `localhost` work, you need to specify an additional docker option:
```
docker run --net="host" -e "DEVICE_ID=$(astartectl utils device-id generate-random)" -e "PAIRING_URL=http://localhost:4003" -e "REALM=<realm name>" -e "PAIRING_JWT=$(astartectl utils gen-jwt pairing -k <realm private key>)" -e astarte/astarte-stream-qt5-test:1.0.4
```
Both commands will generate a random datastream from a brand new, random Device ID.

You can have the same Device ID send longer streams of data by saving the container's persistency through a Docker Volume. If you wish to do so, simply add `-v /persistency:<your persistency path>` to your `docker run` invocation.

You can refer to the previous section for more details on which variables (function, interface, etc...) can be passed to the container.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Add server commands to pause/resume/change streaming params.

## [1.0.4] - 2022-10-25

## [1.0.3] - 2022-07-08

## [1.0.2] - 2022-04-01

## [1.0.1] - 2021-12-20
### Changed
- Add support for the new naming introduced in the
  [Qt5 SDK](https://github.com/astarte-platform/astarte-device-sdk-qt5/pull/28).
- Deprecate `AGENT_KEY` env variable in docker, use `PAIRING_JWT` instead.
- Deprecate `PAIRING_HOST` env variable in docker, use `PAIRING_URL` instead.

## [1.0.0] - 2021-06-30

## [1.0.0-rc.0] - 2021-05-10

## [1.0.0-beta.2] - 2021-03-26

## [1.0.0-beta.1] - 2021-02-16
### Changed
- Bump interface org.astarte-platform.genericsensors.Values version to 1.0.

### Added
- Allow setting the MQTT keepalive.

## [0.11.4] - 2021-01-27

## [0.11.3] - 2020-09-24

## [0.11.2] - 2020-08-14

## [0.11.1] - 2020-05-18

## [0.11.0] - 2020-04-13

## [0.11.0-rc.1] - 2020-03-26

## [0.11.0-rc.0] - 2020-02-27
### Changed
- Use org.astarte-platform.genericsensors.Values interface (and remove org.astarteplatform.Values).

## [0.10.2] - 2020-02-14
### Changed
- Use Debian Buster as base image.

## [0.10.1] - 2019-10-02

## [0.10.0] - 2019-04-17

## [0.10.0-rc.0] - 2019-04-03

## [0.10.0-beta.3] - 2018-12-19

## [0.10.0-beta.2] - 2018-10-19

## [0.10.0-beta.1] - 2018-08-27
### Added
- First Astarte release.


docker-entrypoint.sh
#!/bin/bash

set -e

# Support old JWT naming, print warning
if [[ -v AGENT_KEY && ! -v PAIRING_JWT ]]; then
    echo "AGENT_KEY is deprecated and will be removed in a future release."
    echo "Start using PAIRING_JWT instead."
    PAIRING_JWT=$AGENT_KEY
fi

# Support old PAIRING_HOST, print warning
if [[ -v PAIRING_HOST && ! -v PAIRING_URL ]]; then
    echo "PAIRING_HOST is deprecated and will be removed in a future release."
    echo "Start using PAIRING_URL instead."
    PAIRING_URL=$PAIRING_HOST
fi

# Configure the device directory
mkdir astarte-device-$DEVICE_ID-conf
TRANSPORT_CONF_PATH="astarte-device-${DEVICE_ID}-conf/transport-astarte.conf"
cat > $TRANSPORT_CONF_PATH <<EOF
[AstarteTransport]
pairingUrl=$PAIRING_URL
realm=$REALM
persistencyDir=$PERSISTENCY_DIR
ignoreSslErrors=$IGNORE_SSL_ERRORS
keepAliveSeconds=$KEEPALIVE
EOF

# Check if CREDENTIALS_SECRET and PAIRING_JWT are both set, print warning
if [[ -v CREDENTIALS_SECRET && -v PAIRING_JWT ]]; then
    echo "CREDENTIALS_SECRET and PAIRING_JWT both set, CREDENTIALS_SECRET will be preferred."
    echo "credentialsSecret=$CREDENTIALS_SECRET" >> $TRANSPORT_CONF_PATH
elif [[ -v CREDENTIALS_SECRET ]]; then
    echo "credentialsSecret=$CREDENTIALS_SECRET" >> $TRANSPORT_CONF_PATH
else
    echo "pairingJwt=$PAIRING_JWT" >> $TRANSPORT_CONF_PATH
fi

exec "$@"


AstarteStreamQt5Test.h
/*
 * This file is part of Astarte.
 *
 * Copyright 2018 Ispirata Srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef ASTARTESTREAMQT5TEST_H
#define ASTARTESTREAMQT5TEST_H

#include "AstarteStreamTelemetry.h"

#include <QtCore/QObject>

namespace Hemera {
class Operation;
}

class QTimer;

class AstarteDeviceSDK;

class AstarteStreamQt5Test : public QObject
{
    Q_OBJECT

public:
    AstarteStreamQt5Test(const QByteArray &interface, const QByteArray &path, const QString &function, const QString &device, int interval = 1, double scale = 1, QObject *parent = nullptr);
    ~AstarteStreamQt5Test();

private slots:
    void checkInitResult(Hemera::Operation *op);
    void handleIncomingData(const QByteArray &interface, const QByteArray &path, const QVariant &value);

private:
    AstarteDeviceSDK *m_sdk;
    QByteArray m_interface;
    QByteArray m_path;
    bool m_ready;
    QMap<AstarteStreamTelemetry::TelemetryFunction, AstarteStreamTelemetry*> m_telemetries;
};

#endif


AstarteStreamQt5Test.cpp
/*
 * This file is part of Astarte.
 *
 * Copyright 2018 Ispirata Srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AstarteStreamQt5Test.h"

#include <AstarteDeviceSDK.h>

#include <HemeraCore/Operation>

#include <QtCore/QDir>
#include <QtCore/QTimer>

#define TELEMETRY_CONFIG_INTERFACE "org.astarte-platform.config.Telemetry"

AstarteStreamQt5Test::AstarteStreamQt5Test(const QByteArray &interface, const QByteArray &path,
    const QString &function, const QString &device, int interval, double scale, QObject *parent)
    : QObject(parent)
    , m_interface(interface)
    , m_path(path)
    , m_ready(false)
{

    m_sdk = new AstarteDeviceSDK(QDir::currentPath()
            + QStringLiteral("/astarte-device-%1-conf/transport-astarte.conf").arg(device),
        QDir::currentPath() + QStringLiteral("/interfaces"), device.toLatin1());
    AstarteStreamTelemetry::TelemetryFunction telemetryFunction
        = AstarteStreamTelemetry::getTelemetryFunction(function.toUtf8());
    AstarteStreamTelemetry *telemetry
        = new AstarteStreamTelemetry(m_sdk, interface, path, telemetryFunction, interval, scale);
    m_telemetries.insert(telemetryFunction, telemetry);

    connect(
        m_sdk->init(), &Hemera::Operation::finished, this, &AstarteStreamQt5Test::checkInitResult);
    connect(
        m_sdk, &AstarteDeviceSDK::dataReceived, this, &AstarteStreamQt5Test::handleIncomingData);
}

AstarteStreamQt5Test::~AstarteStreamQt5Test() {}

void AstarteStreamQt5Test::checkInitResult(Hemera::Operation *op)
{
    if (op->isError()) {
        qWarning() << "AstarteStreamQt5Test init error: " << op->errorName() << op->errorMessage();

    } else {
        m_ready = true;
        if (!m_telemetries.isEmpty()) {
            m_telemetries.first()->start();
        }
    }
}

void AstarteStreamQt5Test::handleIncomingData(
    const QByteArray &interface, const QByteArray &path, const QVariant &value)
{
    qDebug() << "Received data, interface: " << interface << "path: " << path
             << ", value: " << value << ", Qt type name: " << value.typeName();

    if (interface.compare(TELEMETRY_CONFIG_INTERFACE) == 0) {
        QList<QByteArray> paths = path.split('/');

        if (paths.isEmpty() || paths.length() <= 3) {
            return;
        }

        QByteArray functionName = paths[2];
        AstarteStreamTelemetry::TelemetryFunction telemetryFunction
            = AstarteStreamTelemetry::getTelemetryFunction(functionName);

        AstarteStreamTelemetry *telemetry;

        if (m_telemetries.contains(telemetryFunction)) {
            telemetry = m_telemetries.value(telemetryFunction);
        } else {
            telemetry = new AstarteStreamTelemetry(m_sdk, m_interface, m_path, telemetryFunction);
            m_telemetries.insert(telemetryFunction, telemetry);
        }

        QByteArray endpoint = paths[3];
        if (endpoint.compare("periodMS") == 0) {
            telemetry->setInterval(value.toInt());
        } else if (endpoint.compare("scale") == 0) {
            telemetry->setScale(value.toDouble());
        }

        if (endpoint.compare("enable") == 0) {
            telemetry->setEnable(value.toBool());
        } else {
            telemetry->restart();
        }
    }
}


AstarteStreamTelemetry.h
/*
 * This file is part of Astarte.
 *
 * Copyright 2023 SECO Mind Srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef ASTARTESTREAMTELEMETRY_H
#define ASTARTESTREAMTELEMETRY_H

#include <AstarteDeviceSDK.h>

#include <QtCore/QObject>
#include <QtCore/QTimer>

class AstarteStreamTelemetry : public QObject
{
    Q_OBJECT

public:
    enum class TelemetryFunction
    {
        Default = 0,
        Sin = 1,
        NoiseSin = 2,
        RandomSpikesSin = 3,
        X = 4,
        Saw = 5,
        Rect = 6,
        Sinc = 7,
        Random = 8
    };
    Q_ENUM(TelemetryFunction);

    AstarteStreamTelemetry(AstarteDeviceSDK *sdk, const QByteArray &interface,
        const QByteArray &path, AstarteStreamTelemetry::TelemetryFunction function, int interval = 1000, double scale = 1,
        QObject *parent = nullptr);
    ~AstarteStreamTelemetry();

    void start();
    void stop();
    void restart();
    void setInterval(int interval);
    void setScale(double scale);
    void setEnable(bool enable);

    static TelemetryFunction getTelemetryFunction(const QByteArray &functionName);

private slots:
    void sendValues();

private:
    AstarteDeviceSDK *m_sdk;
    QByteArray m_interface;
    QByteArray m_path;
    TelemetryFunction m_function;
    QTimer *m_updateTimer;
    double m_scale;
    double m_xValue;

    static int randomInterval();
};

#endif // ASTARTESTREAMTELEMETRY_H


AstarteStreamTelemetry.cpp
/*
 * This file is part of Edgehog.
 *
 * Copyright 2023 SECO Mind Srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AstarteStreamTelemetry.h"

#include <QtCore/QDir>

#include <climits>
#include <cmath>

AstarteStreamTelemetry::AstarteStreamTelemetry(AstarteDeviceSDK *sdk, const QByteArray &interface,
    const QByteArray &path, TelemetryFunction function, int interval, double scale, QObject *parent)
    : QObject(parent)
    , m_sdk(sdk)
    , m_interface(interface)
    , m_path(path)
    , m_function(function)
    , m_updateTimer(new QTimer(this))
    , m_scale(scale)
    , m_xValue(0)

{
    this->setInterval(interval);
    connect(m_updateTimer, &QTimer::timeout, this, &AstarteStreamTelemetry::sendValues);
}

AstarteStreamTelemetry::~AstarteStreamTelemetry() {}

void AstarteStreamTelemetry::start()
{
    if (!m_updateTimer->isActive()) {
        m_updateTimer->start();
    }
}

void AstarteStreamTelemetry::restart()
{
    if (m_updateTimer->isActive()) {
        m_updateTimer->stop();
    }

    m_updateTimer->start();
}

void AstarteStreamTelemetry::stop()
{
    if (m_updateTimer->isActive()) {
        m_updateTimer->stop();
    }
}

void AstarteStreamTelemetry::setInterval(int interval)
{
    if (interval > 0) {
        m_updateTimer->setInterval(interval);
    } else {
        qsrand(QDateTime::currentMSecsSinceEpoch() % UINT_MAX);
        m_updateTimer->setInterval(randomInterval());
        m_updateTimer->setSingleShot(true);
    }
}

void AstarteStreamTelemetry::setScale(double scale)
{
    m_scale = scale;
    this->restart();
}

void AstarteStreamTelemetry::setEnable(bool enable)
{
    if (enable) {
        this->start();
    } else {
        this->stop();
    }
}

void AstarteStreamTelemetry::sendValues()
{
    double x = m_xValue;
    double value;

    switch (m_function) {
        case TelemetryFunction::Default: {
            value = (4 / M_PI) * sin(x) + (4 / 3) * M_PI * sin(3 * x) + (4 / 5) * M_PI * sin(5 * x)
                + (4 / 7) * M_PI * sin(7 * x);
        } break;
        case TelemetryFunction::Sin: {
            value = sin(x);
        } break;
        case TelemetryFunction::NoiseSin: {
            value = sin(x) + ((double) qrand()) / ((double) RAND_MAX) * 0.2;
        } break;
        case TelemetryFunction::RandomSpikesSin: {
            value = sin(x) + (((double) qrand()) / ((double) RAND_MAX) * 0.1)
                + ((((double) qrand()) / ((double) RAND_MAX)) > 0.999 ? 100 : 0);
        } break;
        case TelemetryFunction::X: {
            value = x;
        } break;
        case TelemetryFunction::Saw: {
            value = (fmod(x, 2 * M_PI) - M_PI) / M_PI;
        } break;
        case TelemetryFunction::Rect: {
            value = (fmod(x, 2 * M_PI) - M_PI) > 0 ? 1 : 0;
        } break;
        case TelemetryFunction::Sinc: {
            double t = fmod(x, 10 * 2 * M_PI) - M_PI * 10;
            value = (t != 0) ? (sin(t) / t) : 1;
        } break;
        case TelemetryFunction::Random: {
            value = ((double) qrand()) / ((double) RAND_MAX);
        } break;
    }

    qInfo() <<"functionId: " << m_function <<" " << x << ": " << QString::number(value, 'f', 5);

    m_sdk->sendData(m_interface, m_path, value, QDateTime::currentDateTime());

    if (m_updateTimer->isSingleShot()) {
        m_updateTimer->setInterval(randomInterval());
        m_updateTimer->start();
    }

    m_xValue += M_PI * 2 * ((double) m_updateTimer->interval()) * m_scale;
}

int AstarteStreamTelemetry::randomInterval()
{
    return (qrand() % 600) * 1000 + (qrand() % 1000);
}

AstarteStreamTelemetry::TelemetryFunction AstarteStreamTelemetry::getTelemetryFunction(
    const QByteArray &functionName)
{
    if (functionName == QStringLiteral("sin")) {
        return AstarteStreamTelemetry::TelemetryFunction::Sin;
    } else if (functionName == QStringLiteral("noisesin")) {
        return AstarteStreamTelemetry::TelemetryFunction::NoiseSin;
    } else if (functionName == QStringLiteral("randomspikessin")) {
        return AstarteStreamTelemetry::TelemetryFunction::RandomSpikesSin;
    } else if (functionName == QStringLiteral("x")) {
        return AstarteStreamTelemetry::TelemetryFunction::X;
    } else if (functionName == QStringLiteral("saw")) {
        return AstarteStreamTelemetry::TelemetryFunction::Saw;
    } else if (functionName == QStringLiteral("rect")) {
        return AstarteStreamTelemetry::TelemetryFunction::Rect;
    } else if (functionName == QStringLiteral("sinc")) {
        return AstarteStreamTelemetry::TelemetryFunction::Sinc;
    } else if (functionName == QStringLiteral("random")) {
        return AstarteStreamTelemetry::TelemetryFunction::Random;
    } else {
        return AstarteStreamTelemetry::TelemetryFunction::Default;
    }
}


main.cpp
/*
 * This file is part of Astarte.
 *
 * Copyright 2018 Ispirata Srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AstarteStreamQt5Test.h"

#include <QtCore/QCommandLineParser>
#include <QtCore/QCoreApplication>
#include <QtCore/QDebug>
#include <QtCore/QObject>

#include <HemeraCore/Operation>

int main(int argc, char *argv[])
{
    QCoreApplication app(argc, argv);

    app.setApplicationName(QObject::tr("Astarte Qt5 Stream Test"));
    app.setOrganizationDomain(QStringLiteral("astarte-platform.org"));
    app.setOrganizationName(QStringLiteral("Astarte"));

    QCommandLineParser parser;
    parser.setApplicationDescription(QObject::tr("Astarte stream test."));
    parser.addVersionOption();
    parser.addHelpOption();

    parser.addOptions({
        {
            QStringList{QStringLiteral("f"), QStringLiteral("function")},
            QObject::tr("Function that will be generated."),
            QStringLiteral("function"),
            QStringLiteral("random")
        },
        {
            QStringList{QStringLiteral("s"), QStringLiteral("scale")},
            QObject::tr("Scale that will be used."),
            QStringLiteral("scale"),
            QStringLiteral("0.0001")
        },
        {
            QStringList{QStringLiteral("d"), QStringLiteral("device")},
            QObject::tr("Device ID that will be used."),
            QStringLiteral("device")
        },
        {
            QStringList{QStringLiteral("i"), QStringLiteral("interval")},
            QObject::tr("Interval in ms between two samples."),
            QStringLiteral("interval"),
            QStringLiteral("1000")
        },
        {
            QStringList{QStringLiteral("n"), QStringLiteral("interface")},
            QObject::tr("Target interface"),
            QStringLiteral("interface"),
            QStringLiteral("org.astarte-platform.genericsensors.Values")
        },
        {
            QStringList{QStringLiteral("p"), QStringLiteral("path")},
            QObject::tr("Target path."),
            QStringLiteral("path"),
            QStringLiteral("/streamTest/value")
        }
    });

    parser.process(app);

    bool ok;
    QByteArray interface = parser.value(QStringLiteral("interface")).toLatin1();
    QByteArray path = parser.value(QStringLiteral("path")).toLatin1();
    QString function = parser.value(QStringLiteral("function"));
    QString deviceId = parser.value(QStringLiteral("device"));

    QByteArray decodedDeviceId = QByteArray::fromBase64(deviceId.toLatin1(), QByteArray::Base64UrlEncoding);
    if (decodedDeviceId.count() != 16) {
        qWarning() << deviceId << " doesn't seem to be a valid device ID! Make sure to use an URL encoded base64 128 bits ID.";
        return EXIT_FAILURE;
    }

    int interval;
    if (parser.value(QStringLiteral("interval")) == QStringLiteral("random")) {
        interval = -1;
    } else {
        interval = parser.value(QStringLiteral("interval")).toInt(&ok);
        if (!ok) {
            qWarning() << "Invalid interval.";
            return EXIT_FAILURE;
        }
    }
    double scale = parser.value(QStringLiteral("scale")).toDouble(&ok);
    if (!ok) {
        qWarning() << "Invalid scale.";
        return EXIT_FAILURE;
    }

    AstarteStreamQt5Test astarte(interface, path, function, deviceId, interval, scale);
    return app.exec();
}
